<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Откройте-и-изучите-файл" data-toc-modified-id="Откройте-и-изучите-файл-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откройте и изучите файл</a></span><ul class="toc-item"><li><span><a href="#Выводы" data-toc-modified-id="Выводы-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li><li><span><a href="#Разбейте-данные-на-выборки" data-toc-modified-id="Разбейте-данные-на-выборки-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разбейте данные на выборки</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Исследуйте-модели" data-toc-modified-id="Исследуйте-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследуйте модели</a></span><ul class="toc-item"><li><span><a href="#Модель-решающее-дерево" data-toc-modified-id="Модель-решающее-дерево-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Модель решающее дерево</a></span></li><li><span><a href="#Модель-случайный-лес" data-toc-modified-id="Модель-случайный-лес-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Модель случайный лес</a></span></li><li><span><a href="#Модель-логистической-регрессии" data-toc-modified-id="Модель-логистической-регрессии-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Модель логистической регрессии</a></span></li><li><span><a href="#Сравнение-моделей-между-собой" data-toc-modified-id="Сравнение-моделей-между-собой-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Сравнение моделей между собой</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li><li><span><a href="#Проверьте-модель-на-тестовой-выборке" data-toc-modified-id="Проверьте-модель-на-тестовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверьте модель на тестовой выборке</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#(бонус)-Проверьте-модели-на-адекватность" data-toc-modified-id="(бонус)-Проверьте-модели-на-адекватность-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>(бонус) Проверьте модели на адекватность</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Рекомендация тарифов

**Цель проекта**: построить модель для задачи классификации, которая выберет подходящий новый тариф («Смарт» или «Ультра») для пользователей архивных тарифов. Значение *accuracy* для модели на тестовой выборке должно быть не менее 0,75.

**Исходные данные**: архивные данные о поведении клиентов, которые уже перешли на новые тарифы. Исходные данные не нуждаются в предобработке, так как уже были обработаны в рамках проекта "Определение перспективного тарифа для телеком компании".

**Описание данных:**
* сalls — количество звонков,
* minutes — суммарная длительность звонков в минутах,
* messages — количество sms-сообщений,
* mb_used — израсходованный интернет-трафик в Мб,
* is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0)

## Откройте и изучите файл

In [1]:
# загружаем необходимые библиотеки
import pandas as pd
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
# загружаем файл с исходными данными
df = pd.read_csv("/datasets/users_behavior.csv")
# выводим основную информацию об исходных данных
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
calls       3214 non-null float64
minutes     3214 non-null float64
messages    3214 non-null float64
mb_used     3214 non-null float64
is_ultra    3214 non-null int64
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [2]:
# выводим дополнительно основную описательную статистику
df.describe()

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

### Выводы

1. Файл с исходными данными (размер 3214х5) загружен и преобразован в датафрейм.
2. Пропусков в исходных данных нет - исходные данные уже прошли предобработку.

## Разбейте данные на выборки

In [3]:
# разделим датафрей на две части: с целевым признаком и без
features = df.drop(["is_ultra"], axis=1)
target = df["is_ultra"]

In [4]:
# разделим исходные данные в отношении 3:1:1 на три выборки: обучающую, валидационную и тестовую
features_train, features_train_valid, target_train, target_train_valid = train_test_split(
    features, target, test_size=0.4, random_state=56786)
features_test, features_valid, target_test, target_valid = train_test_split(
    features_train_valid, target_train_valid, test_size=0.50, random_state=56786)
print("Размер обучающей выборки равен:", features_train.shape[0], "строк.")
print("Размер валидационной выборки равен:", features_valid.shape[0], "строк.")
print("Размер тестовой выборки равен:", features_test.shape[0], "строк.")

Размер обучающей выборки равен: 1928 строк.
Размер валидационной выборки равен: 643 строк.
Размер тестовой выборки равен: 643 строк.


### Вывод

Исходные данные разделены в отношении 3:1:1 на три выборки: обучающую, валидационную и тестовую.

## Исследуйте модели

Исследуем три модели: решающее дерево, случайный лес и логистическая регрессия и сравним полученные результаты.

### Модель решающее дерево

In [5]:
# создаем списки - будущие столбы датафрейма, где сохраним результаты вычислений
column_1 = []
column_2 = []
# подсчитаем точность (accuracy) для обученных моделей с различными гиперпараметрами
for depth in range(1, 6):
    model = DecisionTreeClassifier(random_state=56789, max_depth=depth)
    model.fit(features_train, target_train)
    result = model.score(features_valid, target_valid)
    column_1.append(depth)
    column_2.append(result)

In [6]:
# создадим датафрейм с результатами расчета точности обученных моделей
tree_accuracy_valid = pd.DataFrame({"max_depth": column_1, "accuracy": column_2}).sort_values(by="accuracy", ascending=False)
tree_accuracy_valid["accuracy"] = tree_accuracy_valid["accuracy"].round(2)
print(tree_accuracy_valid)

   max_depth  accuracy
4          5      0.79
2          3      0.78
3          4      0.78
1          2      0.77
0          1      0.74


Как видно из таблицы, наибольшая точность соответствует модели решающего дерева с гиперпараметром max_depth=3

### Модель случайный лес

In [7]:
# создаем списки - будущие столбы датафрейма, где сохраним результаты вычислений
column_1 = []
column_2 = []
column_3 = []
# подсчитаем точность (accuracy) для обученных моделей с различными гиперпараметрами
for est in range(2, 11, 2):
    for depth in range(1, 6):
        model = RandomForestClassifier(random_state=56789, n_estimators=est, max_depth=depth)
        model.fit(features_train, target_train)
        result = model.score(features_valid, target_valid)
        column_1.append(est)
        column_2.append(depth)
        column_3.append(result)

In [8]:
# создадим датафрейм с результатами расчета точности обученных моделей
forest_accuracy_valid = pd.DataFrame({"n_estimators": column_1, "max_depth": column_2, "accuracy": column_3})
forest_accuracy_valid.sort_values(by="accuracy", ascending=False, inplace=True)
forest_accuracy_valid["accuracy"] = forest_accuracy_valid["accuracy"].round(3)
forest_accuracy_valid

n_estimators  max_depth  accuracy
14             6          5     0.795
24            10          5     0.793
19             8          5     0.793
23            10          4     0.792
9              4          5     0.788
18             8          4     0.788
13             6          4     0.788
2              2          3     0.787
12             6          3     0.784
4              2          5     0.784
3              2          4     0.784
8              4          4     0.781
7              4          3     0.774
11             6          2     0.773
21            10          2     0.770
16             8          2     0.767
17             8          3     0.767
22            10          3     0.765
1              2          2     0.760
6              4          2     0.760
0              2          1     0.748
15             8          1     0.747
5              4          1     0.747
20            10          1     0.747
10             6          1     0.743

Максимальную точность имеет модель случайного леса при следующих гиперпараметрах: n_estimators = 6, max_depth = 5.

### Модель логистической регрессии

In [9]:
model = LogisticRegression(random_state=56789)
model.fit(features_train, target_train)
score_logistic_regression = model.score(features_valid, target_valid)
print("Accuracy для модели LogisticRegression:", score_logistic_regression)

Accuracy для модели LogisticRegression: 0.6951788491446346


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Логистическая регрессия имеет довольно малую точность - 0,695.

### Сравнение моделей между собой

In [10]:
# построим сводную таблицу с результатами расчета точности для разных моделей
model_list = ["Decision_Tree", "Random_Forest", "Logistic_Regression"]
accuracy_list = [tree_accuracy_valid["accuracy"].max(), forest_accuracy_valid["accuracy"].max(), score_logistic_regression]
max_depth_list = [tree_accuracy_valid.loc[tree_accuracy_valid.idxmax()["accuracy"], "max_depth"], 
                  forest_accuracy_valid.loc[forest_accuracy_valid.idxmax()["accuracy"], "max_depth"],
                  None]
n_estimators_list = [None, 
                     forest_accuracy_valid.loc[forest_accuracy_valid.idxmax()["accuracy"], "n_estimators"], 
                     None]
models_accuracy = pd.DataFrame({"model": model_list, 
                                "max_accuracy": accuracy_list, 
                                "max_depth": max_depth_list, 
                                "n_estimators": n_estimators_list})
models_accuracy

model  max_accuracy  max_depth  n_estimators
0        Decision_Tree      0.790000        5.0           NaN
1        Random_Forest      0.795000        5.0           6.0
2  Logistic_Regression      0.695179        NaN           NaN

Как видно из таблицы, наибольшую точность, равную 0,795, имеет модель случайного леса при следующих гиперпараметрах: n_estimators = 6, max_depth = 5.

### Выводы

1. Было рассмотрено три различных модели: модель решающего дерева, модель случайного леса и модель логистической регрессии.
2. Из всех рассмотренных моделей наибольшую точность, равную 0.795, имеет модель случайного леса при следующих гиперпараметрах: n_estimators = 6, max_depth = 5. Будем считать данную модель наилучшей и проверим ее на тестовой выборке.

## Проверьте модель на тестовой выборке

In [11]:
# проверим лучшую модель (с наибольшей точностью) на тестовой выборке
model = RandomForestClassifier(random_state=56789, n_estimators = 6, max_depth = 5)
model.fit(features_train, target_train)
best_accuracy_test = model.score(features_test, target_test)
best_accuracy_valid = model.score(features_valid, target_valid)
print("Точность лучшей модели на основе тестовой выборки:", best_accuracy_test)
print("Точность лучшей модели на основе валидационной выборки:", best_accuracy_valid)

Точность лучшей модели на основе тестовой выборки: 0.807153965785381
Точность лучшей модели на основе валидационной выборки: 0.7947122861586314


### Вывод

Точность лучшей модели на тестовой выборке равняется 0,807, что больше точности для той же модели на валидационной выборке, равной 0,795.

## (бонус) Проверьте модели на адекватность

Посмотрим как распределяются доли пользователей с различными тарифами Смарт и Ультра.

In [12]:
df["is_ultra"].value_counts() / len(df)

0    0.693528
1    0.306472
Name: is_ultra, dtype: float64

Если мы примем случайную модель, что все тарифы в выборке - это тарифы Смарт, тогда точность такой модели составит 69,3%. Точность нашей модели существенно выше - около 80%. Таким образом, ее можно считать адекватной.

### Вывод

Полученную модель случайного леса на основе сравнения со случайной моделью следует считать адекватной.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
